### PR TITLE
GEOMESA-1243 Cleaning up kafka connections on data store dispose, upd…

### PR DESCRIPTION
**Pull request by elahrvivaz**
…ating synchronization in BucketIndex

Signed-off-by: Emilio Lahr-Vivaz <elahrvivaz@ccri.com>

### CODE REVIEW

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/index/BucketIndex.scala [hunk 1]
```diff
@@ -20,20 +22,22 @@ class BucketIndex[T](xBuckets: Int = 360,
...
+  // create the buckets up front to avoid having to synchronize the whole array
+  // we use a concurrentHashSet, which gives us iterators that aren't affected by modifications to the backing set
+  private def newSet = Collections.newSetFromMap(new ConcurrentHashMap[T, java.lang.Boolean])
+  private val buckets = Array.fill(xBuckets, yBuckets)(newSet)
 
   override def insert(envelope: Envelope, item: T): Unit = {
     val (i, j) = getBucket(envelope)
     val bucket = buckets(i)(j)
-    bucket.synchronized(bucket.add(item))
+    bucket.add(item)
```
Comment (jnh5y) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/index/BucketIndex.scala]: NICE!  Does this improve performance?

Comment (elahrvivaz) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/index/BucketIndex.scala]: i'm guessing the main performance boost will be from line 65. Thanks for the reminder about ConcurrentHashMaps, btw :)

**Comment by jnh5y** [comment]
Looks good to me.  @mmatz-ccri @ddseapy can you guys weigh in?  @Afterglow375 @jahhulbert-ccri thoughts?

**Comment by elahrvivaz** [comment]
this doesn't have to go in for 1.2.3. probably needs some testing.

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/index/BucketIndex.scala [hunk 2]
```diff
@@ -20,20 +22,22 @@ class BucketIndex[T](xBuckets: Int = 360,
                      yBuckets: Int = 180,
                      extents: Envelope = new Envelope(-180.0, 180.0, -90.0, 90.0)) extends SpatialIndex[T] {
 
-  // create the buckets upfront to avoid having to synchronize the whole array
   private val gridSnap = new GridSnap(extents, xBuckets, yBuckets)
-  private val buckets = Array.fill(xBuckets, yBuckets)(mutable.HashSet.empty[T])
+  // create the buckets up front to avoid having to synchronize the whole array
+  // we use a concurrentHashSet, which gives us iterators that aren't affected by modifications to the backing set
```
Comment (mmatz-ccri) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/index/BucketIndex.scala]: ConcurrentHashMap?

**Comment by mmatz-ccri** [comment]
Took a quick look and seems fine.

**Pull request merged**